Objective-C/Internal/CBLDictionaryObject.m

```diff
@@ -0,0 +1,402 @@
+//
+//  CBLDictionary.m
+//  CouchbaseLite
+//
+//  Created by Pasin Suriyentrakorn on 4/12/17.
+//  Copyright © 2017 Couchbase. All rights reserved.
+//
+
+#import "CBLDictionaryObject.h"
+#import "CBLBlob.h"
+#import "CBLDictionaryData.h"
+#import "CBLDocumentInternal.h"
+#import "CBLJSON.h"
+#import "CBLReadOnlyDocumentInternal.h"
+
+@implementation CBLDictionaryObject {
+    NSMutableDictionary<NSString*, id>* _dict;
+    NSMapTable* _changeListeners;
+    BOOL _changed;
+}
+
+
+- (instancetype) initWithData: (CBLDictionaryData*)data {
+    self = [super initWithData: data];
+    if (self) {
+        _dict = [NSMutableDictionary dictionary];
+    }
+    return self;
+}
+
+
+#pragma mark - GETTER
+
+
+- (NSDictionary<NSString*, id>*) properties {
+    if (self.data.data) {
+        // Bring the backing data up:
+        NSDictionary* backingData = [super properties];
+        [backingData enumerateKeysAndObjectsUsingBlock: ^(id key, id value, BOOL *stop) {
+            if (!_dict[key])
+                _dict[key] = [self prepareValue: value];
+        }];
+        // Detach backing data as all backing data has been brought up:
+        self.data = [[CBLDictionaryData alloc] init]; // EMPTY
+    }
+    
+    // Filter out keys that have NSNull value to make the result
+    // the same across platforms (Java and .NET do not have NSNull
+    // equivalent value):
+    NSMutableDictionary* result = [_dict mutableCopy];
+    for (NSString* key in _dict) {
+        if (result[key] == [NSNull null])
+            result[key] = nil;
+    }
+    return result;
+}
+
+
+- (nullable id) objectForKey: (NSString*)key {
+    id value = _dict[key];
+    if (!value) {
+        value = [super objectForKey: key];
+        if ([value isKindOfClass: [CBLReadOnlySubdocument class]]) {
+            value = [self convertReadOnlySubdocument: value];
+            [self setValue: value forKey: key isChange: NO];
+        } else if ([value isKindOfClass: [CBLReadOnlyArray class]]) {
+            value = [self convertReadOnlyArray: value];
+            [self setValue: value forKey: key isChange: NO];
+        }
+    } else if (value == [NSNull null])
+        value = nil; // Cross-platform behavior
+    return value;
+}
+
+
+- (BOOL) booleanForKey: (NSString*)key {
+    id value = _dict[key];
+    if (!value)
+        return [super booleanForKey: key];
+    else {
+        if (value == [NSNull null])
+            return NO;
+        else {
+            id n = $castIf(NSNumber, value);
+            return n ? [n boolValue] : YES;
+        }
+    }
+}
+
+
+- (NSInteger) integerForKey: (NSString*)key {
+    id value = _dict[key];
+    if (!value)
+        return [super integerForKey: key];
+    else
+        return [$castIf(NSNumber, value) integerValue];
```

Comment:
Same as explanation above. It's an optimization. Will get consult with Jens on this.